### PR TITLE
[SPARK-13916][SQL] Add a metric to WholeStageCodegen to measure duration.

### PR DESCRIPTION
**Pull request by nongli**
## What changes were proposed in this pull request?

WholeStageCodegen naturally breaks the execution into pipelines that are easier to
measure duration. This is more granular than the task timings (a task can be multiple
pipelines) and is integrated with the web ui.

We currently report total time (across all tasks), min/mask/median to get a sense of how long each is taking.
## How was this patch tested?

Manually tested looking at the web ui.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
   protected LinkedList<InternalRow> currentRows = new LinkedList<>();
   // used when there is no column in output
   protected UnsafeRow unsafeRow = new UnsafeRow(0);
+  private long startTimeMs = System.currentTimeMillis();
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java]: we should use nanoTime. See https://github.com/databricks/scala-style-guide#misc_currentTimeMillis_vs_nanoTime

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 2]
```diff
@@ -300,6 +304,10 @@ case class WholeStageCodegen(child: SparkPlan) extends UnaryNode with CodegenSup
   override def outputPartitioning: Partitioning = child.outputPartitioning
   override def outputOrdering: Seq[SortOrder] = child.outputOrdering
 
+  override private[sql] lazy val metrics = Map(
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: how does this work if there are two WholeStageCodegens in a single stage?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 3]
```diff
@@ -168,15 +186,24 @@ private[sql] object SQLMetrics {
...
     // data size total (min, med, max):
     // 100GB (100MB, 1GB, 10GB)
-    createLongMetric(sc, s"$name total (min, med, max)", StaticsLongSQLMetricParam)
+    createLongMetric(sc, s"$name total (min, med, max)", StatisticsBytesSQLMetricParam)
+  }
+
+  def createTimingMetric(sc: SparkContext, name: String): LongSQLMetric = {
+    // The final result of this metric in physical operator UI may looks like:
+    // duration(min, med, max):
+    // 1s (800msMB, 1s, 2s)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: msMB?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala (hunk 2 above)
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: Doesn't matter. That would be similar to have two projects in a single stage. The metrics are per instance.

**Comment by rxin** [comment]
LGTM

**Comment by rxin** [comment]
Merging in master. Thanks.

**Pull request closed without merging**